doUpsert/index.js

```diff
@@ -56,24 +72,38 @@ module.exports = async function(context, jsonItem) {
           context,
           digitalTwin,
           jsonItem.$sourceId,
-          jsonItem.$relationshipId);
+          jsonItem.$relationshipId,
+      );
     } else { // relationship must be upserted
-      context.log.verbose(`upserting relationship ${jsonItem.$relationshipId}`);
-      await (async () => {
-        context.log.verbose('waiting 100ms');
-        sleep(100);
-        context.log.verbose('calling ADT relationship API');
-        await digitalTwin.upsertRelationship(
-            jsonItem.$sourceId,
-            jsonItem.$relationshipId,
-            jsonItem.relationship)
-            .catch((e) => {
-              context.log.error(`relationship ${jsonItem.$relationshipId}
+      try { // if relationship exists it must be updated
+        await digitalTwin.getRelationship(
+            jsonItem.$sourceId, jsonItem.$relationshipId);
+        context.log.verbose(
+            `updating relationship ${jsonItem.$relationshipId}`);
+        const rlPatch = buildPatch(jsonItem);
+        await digitalTwin.updateRelationship(
+            jsonItem.$sourceId, jsonItem.$relationshipId, rlPatch);
+      } catch (error) {
```

Comment:
instead of try/catch the updateRelationShip function, use the returned value of getRelationship. If exists => update else upsert